File: doc/src/codes/Users_Manual/stm/focus_0.rs
# extern crate autd3;
# extern crate tokio;
# use autd3::prelude::*;
# #[tokio::main]
# async fn main() -> Result<(), Box<dyn std::error::Error>> {
# let mut autd = Controller::builder().add_device(AUTD3::new(Vector3::zeros())).open_with(autd3::link::Nop::builder()).await?;
let center = autd.geometry.center() + Vector3::new(0., 0., 150.0 * MILLIMETER);
let point_num = 200;
let radius = 30.0 * MILLIMETER;
let stm = FocusSTM::new(1.0).add_foci_from_iter((0..point_num).map(|i| {
    let theta = 2.0 * PI * i as float / point_num as float;
    let p = radius * Vector3::new(theta.cos(), theta.sin(), 0.0);
    center + p
}))?;
# Ok(())
# }

File: doc/src/codes/Users_Manual/stm/focus_0.cpp
#include <ranges>

#include "autd3.hpp"

using namespace std::ranges::views;

const autd3::Vector3 center =
    autd.geometry().center() + autd3::Vector3(0, 0, 150);
const auto points_num = 200;
const auto radius = 30.0;
auto stm = autd3::FocusSTM(1).add_foci_from_iter(
    iota(0) | take(points_num) | transform([&](auto i) {
      const auto theta = 2.0 * autd3::pi * static_cast<double>(i) /
                         static_cast<double>(points_num);
      autd3::Vector3 p = center + autd3::Vector3(radius * std::cos(theta),
                                                 radius * std::sin(theta), 0);
      return p;
    }));
